src/google/protobuf/compiler/csharp/csharp_primitive_field.cc

```diff
@@ -89,6 +120,27 @@ void PrimitiveFieldGenerator::GenerateMembers(io::Printer* printer) {
   printer->Print(
     "  }\n"
     "}\n");
+  if (descriptor_->file()->syntax() == FileDescriptor::SYNTAX_PROTO2) {
+    printer->Print(variables_, "/// <summary>Gets whether the \"$descriptor_name$\" field is set</summary>\n");
+    AddPublicMemberAttributes(printer);
+    printer->Print(
+      variables_, 
+      "$access_level$ bool Has$property_name$ {\n"
```

Comment:
Field presence does not mean we need to provide has methods in each message. As C# can have nullable type, users can check if the field is set by themselves. C++ provide has methods because there is no nullable type and is using has bits that user can't access for field presence. It is a waste to produce has methods for each field which seems "same code". I'd prefer to ask user check null directly. 

Same issue for clear methods.

Reply:
Now that I review that code, I realize I shouldn't have used nullables in the first place. Nullables aren't exactly efficient and messages should use bitfields like the Java, Python, and C++ versions.

So I'll retain the Has and Clear methods and reimplement the underlying field presence to use something that doesn't use add four bytes of memory to each field.

Reply:
Change to an efficient way sounds good to me. But four bytes of memory to each field looks like not good enough.  C++'s  hasbit array is one bit for each field:
https://github.com/google/protobuf/blob/f81a3910c07daba7d1f2fd57e328692fc8c480ca/src/google/protobuf/compiler/cpp/cpp_message.cc#L502